### PR TITLE
Combo boxes

### PR DESCRIPTION
**Pull request by ljjh20**
Currently very messy as I am jumping between two different ways of adding the right window widgets, by commenting out other when not used. Slot is in qt_pisces_main_window.py in constructor. Signal is in qt_view_laser_mode.py

### CODE REVIEW

File: widgets/qt_pisces_main_window.py [hunk 1]
```diff
@@ -4,9 +4,10 @@
 from PyQt5.QtGui import *
 import qt_textbox_and_slider
 import qt_custom_decorations
+import qt_view_laser_mode
 
 
-class main_widgetWindow(QMainWindow):
+class MainWidgetWindow(QMainWindow):
```
Comment (AhmetCanSolak) [inline widgets/qt_pisces_main_window.py:10]: Correct naming convention :+1:

File: widgets/qt_pisces_main_window.py [hunk 2]
```diff
@@ -16,80 +17,126 @@ def __init__(self, *args, **kwargs):
...
+        self.button_state = False
+
         self.initUI()
 
         self.show()  # not shown by default
 
     def initUI(self):  # Window properties are set in the initUI() method
+
+        @pyqtSlot()
+        def toggleState():
```
Comment (AhmetCanSolak) [inline widgets/qt_pisces_main_window.py:29]: this method is a closure right now, not a member method `MainWidgetWindow` class. Could you make it a member methed like `initUI()`?

**Review by AhmetCanSolak — CHANGES_REQUESTED** [review]

File: widgets/qt_pisces_main_window.py [hunk 3]
```diff
@@ -2,106 +2,131 @@
...
-        left_scroll_bar.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
-        left_scroll_bar.setWidgetResizable(False)
-        left_scroll_bar.setStyleSheet("background : lightgray;")
-        left_scroll_bar.setWidget(left_widget)
-
-        # add left and right window components to main window (left is a widget due to scroll bar)
-        window_layout.addWidget(left_scroll_bar)
-        window_layout.addLayout(right_window_layout)
+        # Left window
+        self.left_window_layout.setAlignment(Qt.AlignTop)
```
Comment (AhmetCanSolak) [inline widgets/qt_pisces_main_window.py:45]: Left half of the window possibly can be refactored into a separate file as a single widget, what do you think @ljjh20

File: widgets/qt_custom_sliders.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class DoubleClickableSlider(QSlider):
+    def __init__(self, parent):
+        super(QSlider, self).__init__(Qt.Horizontal, parent)
+
+        self.parent = parent
+
+    def mouseDoubleClickEvent(self, event):
+        self.parent.toggle_range_widgets()
```
Comment (AhmetCanSolak) [inline widgets/qt_custom_sliders.py:14]: what if parent doesnt have `toggle_range_widgets` member?

File: widgets/qt_textbox_and_slider.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 from PyQt5.QtCore import *
 from PyQt5.QtGui import *
 from math import log10
+import qt_custom_sliders
 
 
 class InitializeSliderTextB(QWidget):
```
Comment (AhmetCanSolak) [inline widgets/qt_textbox_and_slider.py:9]: maybe we can name this class in a more clear way?

File: widgets/qt_view_laser_mode.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import sys
+from PyQt5.QtWidgets import *
+from PyQt5.QtCore import *
+from PyQt5.QtGui import *
+import qt_custom_decorations
+
+
+class InitializeComboButton(QWidget):
```
Comment (AhmetCanSolak) [inline widgets/qt_view_laser_mode.py:8]: also i believe we can name this class better

File: widgets/qt_view_laser_mode.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    def __init__(self, parent, button_name, add_line_break=False, can_disable=False):
+        super(QWidget, self).__init__(parent)
+
+        self.parent = parent
+        self.button_name = button_name
+        self.add_line_break = add_line_break
+        self.can_disable = can_disable
+
+        if self.can_disable:
+            self.button_state = False
```
Comment (AhmetCanSolak) [inline widgets/qt_view_laser_mode.py:18]: else? `self.button_state` wouldn't be defined? is that intentional?

**Review by AhmetCanSolak — CHANGES_REQUESTED** [review]
dropped few comments, then we can merge this PR in.

File: widgets/qt_view_laser_mode.py (hunk 7 above)
Comment (ljjh20) [reply widgets/qt_view_laser_mode.py:18]: yes, this prevents the view button from being able to disable the widgets.

File: widgets/qt_view_laser_mode.py (hunk 6 above)
Comment (ljjh20) [reply widgets/qt_view_laser_mode.py:8]: I named custom_decorations and custom_sliders with the intention of there being multiple custom classes in each file. I don't think I am quite clear on the one class per file, as it forces a double naming structure that feels a bit clunky. Maybe you can explain a good way to to this when we chat.

File: widgets/qt_pisces_main_window.py [hunk 8]
```diff
@@ -2,106 +2,131 @@
...
+        # initialize layouts
+        self.window_layout = QHBoxLayout()
+        self.left_window_layout = QVBoxLayout()
+        self.right_window_layout = QVBoxLayout()
+        self.label_layout = QHBoxLayout()
+        self.controls_label_layout = QHBoxLayout()
+
+        # initialize right window widgets
+        self.view_window = qt_view_laser_mode.InitializeComboButton(self, "View")
+        self.timelapse_window = qt_view_laser_mode.InitializeComboButton(self, "Timelapse", True, True)
```
[the remark this answers is not in the document]
Comment (ljjh20) [reply widgets/qt_pisces_main_window.py:32]: They need to be accessed by member function to disable widgets. I am not sure how to access them if they are in the initUI

[the remark this answers is not in the document]
Comment (AhmetCanSolak) [reply widgets/qt_pisces_main_window.py:32]: I see. Maybe we should consider dropping initUI methods and initialize everything on widget in `__init__` method? I do that way on many other projects and working well for me so far.

File: widgets/qt_view_laser_mode.py (hunk 6 above)
Comment (AhmetCanSolak) [reply widgets/qt_view_laser_mode.py:8]: sure, we can chat about it.

File: widgets/qt_view_laser_mode.py (hunk 7 above)
Comment (AhmetCanSolak) [reply widgets/qt_view_laser_mode.py:18]: ha, i think we can refactor this into something simpler

**Pull request merged**